src/Microsoft.ComponentDetection.Detectors/pip/ISimplePyPiClient.cs

```diff
@@ -0,0 +1,310 @@
+        }
+
+        var response = await this.RetryPypiRequestAsync(uri, spec);
+
+        var responseContent = await response.Content.ReadAsStringAsync();
+        if (string.IsNullOrEmpty(responseContent))
+        {
+            return pythonProject;
+        }
+
+        try
+        {
+            pythonProject = JsonConvert.DeserializeObject<SimplePypiProject>(responseContent);
+        }
+        catch (Exception e)
+        {
+            this.logger.LogError(
+                    e,
+                    "Unable to deserialize simple pypi project. This is possibly because the server responded with an unexpected content type. Spec Name = {SpecName}",
+                    spec.Name);
+            return new SimplePypiProject();
+        }
+
+        // The `first - wins` response accepted into the cache. This might be different from the input if another caller wins the race.
+        return await this.cachedSimplePyPiProjects.GetOrCreateAsync(uri, cacheEntry =>
+        {
+            cacheEntry.SlidingExpiration = TimeSpan.FromSeconds(CACHEINTERVALSECONDS); // This entry will expire after CACHEINTERVALSECONDS seconds from last use
+            cacheEntry.Size = 1; // Specify a size of 1 so a set number of entries can always be in the cache
+            return Task.FromResult(pythonProject);
+        });
+    }
+
+    /// <summary>
+    /// On the initial caching attempt, see if the user specified an override for
+    /// PyPiMaxCacheEntries and recreate the cache if needed.
+    /// </summary>
+    private MemoryCache InitializeNonDefaultMemoryCache(MemoryCache cache)
+    {
+        var maxEntriesVariable = this.environmentVariableService.GetEnvironmentVariable("PyPiMaxCacheEntries");
+        if (!string.IsNullOrEmpty(maxEntriesVariable) && long.TryParse(maxEntriesVariable, out var maxEntries))
+        {
+            this.logger.LogInformation("Setting ISimplePyPiClient max cache entries to {MaxEntries}", maxEntries);
+            cache = new MemoryCache(new MemoryCacheOptions { SizeLimit = maxEntries });
+        }
+
+        this.checkedMaxEntriesVariable = true;
+        return cache;
+    }
+
+    private async Task<HttpResponseMessage> RetryPypiRequestAsync(Uri uri, PipDependencySpecification spec)
+    {
+        var request = await Policy
+           .HandleResult<HttpResponseMessage>(message =>
+           {
+               // stop retrying if MAXRETRIES was hit!
+               if (message == null)
+               {
+                   return false;
+               }
+
+               var statusCode = (int)message.StatusCode;
+
+               // only retry if server doesn't classify the call as a client error!
+               var isRetryable = statusCode != 300 && statusCode != 406 && (statusCode < 400 || statusCode > 499);
+               return !message.IsSuccessStatusCode && isRetryable;
+           })
+           .WaitAndRetryAsync((int)MAXRETRIES - 1, i => RETRYDELAY, (result, timeSpan, retryCount, context) =>
```

Comment:
Does this API return any `Retry-After` header that we can use instead of a static value?

Reply:
Doesn't seem like it does. I don't see it mentioned in the docs